pkg/controller/upgradeconfig/upgradeconfig_controller.go

```diff
@@ -65,8 +61,10 @@ var _ reconcile.Reconciler = &ReconcileUpgradeConfig{}
 type ReconcileUpgradeConfig struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
-	client client.Client
-	scheme *runtime.Scheme
+	client                   client.Client
+	scheme                   *runtime.Scheme
+	maintenanceClientBuilder func(client client.Client) (maintenance.Maintenance, error)
```

Comment:
Does this make more sense being part to the `ClusterUpgrader` struct?

Reply:
I agree - there is more refactoring needed around this. I'm a little hesitant about getting too drastic with changes as part of this PR. Think it's worth carding this up separately.